libsmi/tools/smixlate.1
.\"
.\" $Id: smixlate.1.in 1676 2004-08-10 10:58:12Z strauss $
.\"
.TH smixlate 1  "June 18, 2006" "IUB" "SMI Tools"
.SH NAME
smixlate \- translate SMI/SPPI identifiers
.SH SYNOPSIS
.B smixlate
[
.B "-Vhm"
] [
.BI "-c " file
] [
.BI "-p " module
] [
.BI "-l " level
]
.I "module(s)"
.SH DESCRIPTION

The \fBsmixlate\fP program is used to translate identifiers and
especially OIDs into a more human readable format.
.SH OPTIONS
.TP
\fB-V, --version\fP
Show the smixlate version and exit.
.TP
\fB-h, --help\fP
Show a help text and exit.
.TP
\fB-r, --recursive\fP
Report errors and warnings also for recursively imported modules. 
.TP
\fB-c \fIfile\fB, --config=\fIfile\fP
Read \fIfile\fP instead of any other (global and user)
configuration file.
.TP
\fB-p \fImodule\fB, --preload=\fImodule\fP
Preload the module \fImodule\fP before reading the main
module(s). This may be helpful if an incomplete main module misses to
import some definitions.
.TP
\fB-l \fIlevel\fB, --level=\fIlevel\fP
Report errors and warnings up to the given severity \fIlevel\fP. See
the smilint(1) manual page for a description of the error levels. The
default error level is 3.
.TP
\fB-a, --all\fP
Replace all OIDs including OID prefixes. Without this option, smixlate
will only translate OIDs with a corresponding notification, scalar,
column, row, or table definition.
.TP
\fB-f, --format\fP
Preserve the input format as much as possible by inserting/removing
white space characters.
.TP
.I module(s)
These are the modules to be loaded for the subsequent translation. If
a module argument represents a path name (identified by containing at
least one dot or slash character), this is assumed to be the exact
file to read. Otherwise, if a module is identified by its plain module
name, it is searched according to libsmi internal rules. See
\fBsmi_config(3)\fP for more details.
.SH "EXAMPLE"
This example translates numeric OIDs in the input text into a more
human readable format.
.nf

  $ echo "what is this oid? 1.3.6.1.2.1.2.2.1.3" | \    
    ./smixlate -l 0 /usr/local/share/mibs/ietf/*
  what is this oid? ifType
  $

.fi
.SH "SEE ALSO"
The 
.BR libsmi (3)
project is documented at
.BR "http://www.ibr.cs.tu-bs.de/projects/libsmi/" "."
.SH "AUTHORS"
(C) 2006-2006 J. Schoenwaelder, International University Bremen, Germany
.br
and contributions by many other people.
.br


qwt/CHANGES
Release 5.1.0
===================

Changes
-------
1) QwtSymbol::copy introduced
   Now it is possible to use derived symbol classes for curves
2) QwtPlotScaleItem introduced
   A new type of plot item for displaying axes on the canvas
3) QwtClipper added
   A collection of clipping algos
4) Using DBL_EPSILON
   This change allows smaller intervals for sliders/dials
5) QwtPanner
   setOrientation() added.
6) QwtPlot
   axisStepSize() added
   clear is virtual now
7) QwtPlotPrintFilter
   PrintCanvasBackground splitted into PrintBackground, PrintFrameWithScales
8) QwtPlotZoomer
   setZoomStack() added
9) Changes for the QwtPolar package
   QwtLegendItemManager introduced
   QwtMagnifier introduced
10)Suffix rules added in qwtconfig.pri for different targets for 
   debug/release builds.

Bug Fixes
---------
1. QwtAbstractScaleDraw::setAbstractScaleDraw
   Reinitialization problem fixed
2. QwtLegendItem
   key event handlers fixed
3. QwtPicker
   solaris-cc compiler problem fixed
4. Inaccurate mapping of scale to widget coordinates fixed
5. QwtPlotCurve::draw
   Updates for Qt 4.3 added
6. QwtPlotLayout
   AlignToCanvas layout calculation fixed
7. QwtPlot::print
   Workaround for a QPen initialization problem, 
   when printing to Pdf, added
8. QwtText
   Layout of rich text documents fixed
9. Designer
   Handling of QwtScaleWidget fixed
10. realtime example
   Qt::WA_PaintOutsidePaintEvent added, ScrollZoomer fixed
11. Several others I have forgotten

Release 5.0.2
===================

Bug Fixes
---------
1. QwtPlotCurve::Xfy curve type fixed
2. Memory leak in QwtLegend fixed
3. Vertical alignment of rich texts fixed
4. Workaround for a Qt4 bug added, that produces horrible performance
   when painting curves with a pen width > 1.
5. Background for the tracker text of QwtPickers fixed.
   Improved (faster + better rendered texts) implementation of
   painting tracker texts, using capabilities of Qt >= 4.3.
6. QwtArrowButton/QwtCounter: workaround for layout bug ( Qt < 4.3 )
   of the Cleanlook style added.
7. A couple of minor fixes

Changes
-------
1. QSvgGenerator added to the bode example

Release 5.0.1
===================

Changes
-------
1.  A couple of problems, when building Qwt fixed.
2.  Displaying Rich Text with Qt 4.x fixed

Release 5.0.0
===================

Platforms
---------
Support of Qt3 and Qt4. Qt2 is not supported any longer.

Key features
------------
1.  Redesign of plot items. Makes it much easier to develop
    individual items.
2.  Redesign of the scale classes. All calculations are
    collected in scale engines, where the application can
    implement it´s own (f.e log2, or date scales). Now it´s
    also possible to have individual and completely irregular scales
3.  Redesign of the QwtText classes. The MathML renderer of
    the Qt4 solutions package is embedded. 
    work for all expressions/situations.
4.  New classes for navigating: QwtPanner, QwtMaginfier
5.  Spectrogram/Contour plots  and other classes for displaying
    raster data added.

Changes
-------
5.0.0 is by far the release with the most changes in the history of Qwt
- too many to make list.


Release 4.2.0/0.4.2
===================

License
--------
A couple of exceptions to the LGPL with the intention to allow static
linking with commercial applications. See COPYING.

Key features: 
-------------
1.  Designer plugin 
2.  Rich Text support ( f.e. E = m * c<sup>2</sup> ) added.
3.  QwtDial class family added (QwtDial, QwtCompass, QwtAnalogClock, ...)
4.  QwtPicker class family added. Includes QwtPlotZoomer, a complete
    implementation of recursive zooming.
5.  Device metrics independent printing of QwtPlot. (QPrinter::HighResolution)
6.  QwtPlot::setCurveBrush(), QwtCurve::setBrush() added. The area
    between curve and baseline will be filled with this brush.
7.  Rotation of axis tick labels added. Very useful for axis with long
    labels like time scales ...
8.  Added a new abstract QwtData class to plot data from almost any type of
    container class.  
9.  QwtDoublePoint, QwtDoubleSize, QwtDoubleRect double counterparts
    for QPoint, QSize, QRect.
10. First steps to support Qtopia. All examples can be compiled and started
    in the qvfb emulator.


Changes:
---------
1.  Rewrite of QwtLegend/QwtLegendItem (no QTable anymore)
2.  Each plot item will be painted, even if one of the axis it is 
    attached to is disabled. (like in all other releases beside 0.4.1)
3.  Code for double buffering moved to a new class QwtPaintBuffer.
    Double buffering can be enabled/disabled now. 
4.  QwtPainter, QwtMetricsMap, QwtLayoutMetrics added
    Hide paint device metrics dependencies.
5.  Layout code rewritten and moved to a new class QwtPlotLayout
    New layout options canvasMargin(), alignCanvasToScales()
6.  QwtPlot: sizeHint() != minimumSizeHint()
9.  Internal plot data are private again. A couple of get methods 
    added instead.
10. canvas repaints triggered by paint events. Enables event filtering
11. QwtPlot::drawCanvasItems added. In opposite to QwtPlot::drawCanvas
    it is used by the printing code too.
12. qwtMax, qwtMin, qwtInt mapped to QMAX, QMIN, qRound from qglobal.h
13. operator= for plot item classes changed.
14. readOnly property added for sliders.
15. valid flag added for QwtDblRange
16. QwtCounter wrap around policy: a counter under- or overflow sets
    focus to the smallest up/down button and disables counting.
    A space bar keypress release event re-enables counting.
17. QwtPushButton added. A class that adds rich text and alignments
    features to QPushButton, like they are used in QLabel
18. Clipped painting code moved from QwtCurve to QwtPainter/QwtRect
19. Canvas cache added to optimize trivial repaints.
20. QwtPlot::drawCurve added for incremental curve data
21. QwtSliderBase, readOnly, isValid added
22. Added filtering of the colors of the title and scales to QwtPrintFilter.
23. Support of QT_NO_CAST_ASII and QT_NO_COMPAT added

libsmi/tools/smixlate.c
/*
 * smixlate.c --
 *
 *      Translate OIDs located in the input stream.
 *
 * Copyright (c) 2006 Juergen Schoenwaelder, International University Bremen.
 *
 * See the file "COPYING" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * @(#) $Id: smilint.c 1867 2004-10-06 13:45:31Z strauss $
 */

#include <config.h>

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <assert.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_WIN_H
#include "win.h"
#endif

#include "smi.h"
#include "shhopt.h"
#include "dstring.h"

static int flags;
static int aFlag = 0;	/* translate all OIDs */
static int fFlag = 0;	/* preserve formatting */

static void translate(dstring_t *token, dstring_t *subst)
{
    SmiNode *smiNode;
    SmiSubid oid[256];
    unsigned int oidlen = 0;
    unsigned int i;
    char *p;

    assert(token && subst);

    dstring_truncate(subst, 0);
    dstring_assign(subst, dstring_str(token));

    for (oidlen = 0, p = strtok(dstring_str(token), ". "); p;
	 oidlen++, p = strtok(NULL, ". ")) {
	oid[oidlen] = strtoul(p, NULL, 0);
    }

    smiNode = smiGetNodeByOID(oidlen, oid);
    if (smiNode &&
	(aFlag
	 || smiNode->nodekind == SMI_NODEKIND_SCALAR
	 || smiNode->nodekind == SMI_NODEKIND_COLUMN
	 || smiNode->nodekind == SMI_NODEKIND_NOTIFICATION
	 || smiNode->nodekind == SMI_NODEKIND_TABLE
	 || smiNode->nodekind == SMI_NODEKIND_ROW)) {
	dstring_assign(subst, smiNode->name);
	for (i = smiNode->oidlen; i < oidlen; i++) {
	    dstring_append_printf(subst, ".%d", oid[i]);
	}
    }
}


static void process(FILE *stream)
{
    int c, space = 0;
    enum { TXT, NUM, NUMDOT, NUMDOTNUM, OID, OIDDOT, EATSPACE } state = TXT;
    dstring_t *token, *subst;

    token = dstring_new();
    subst = dstring_new();
    
    if (! token || ! subst) {
	return;
    }

    /*
     * Shall we require iswhite() or ispunct() before and after the
     * OID?
     *
     * TODO: - translate instance identifier to something meaningful
     *         (e.g. foobar["name",32]) where possible
     *       - generate warnings if instance identifier are incomplete
     *       - provide a reverse translation service (-x) (but this is
     *         more complex since it is unclear how to identify names
     *	     - make the white space magic optional
     */

    while ((c = fgetc(stream)) != EOF) {
	switch (state) {
	case TXT:
	    fputs(dstring_str(token), stdout);
	    dstring_truncate(token, 0);
	    if (isdigit(c) && c >= '0' && c <= '2') {
		dstring_append_char(token, (char) c);
		state = NUM;
	    } else {
		fputc(c, stdout);
		fflush(stdout);
	    }
	    break;
	case NUM:
	    if (isdigit(c)) {
		dstring_append_char(token, (char) c);
	    } else if (c == '.') {
		dstring_append_char(token, (char) c);
		state = NUMDOT;
	    } else {
		dstring_append_char(token, (char) c);
		state = TXT;
	    }
	    break;
	case NUMDOT:
	    if (isdigit(c)) {
		dstring_append_char(token, (char) c);
		state = NUMDOTNUM;
	    } else {
		dstring_append_char(token, (char) c);
		state = TXT;
	    }
	    break;
	case NUMDOTNUM:
	    if (isdigit(c)) {
		dstring_append_char(token, (char) c);
	    } if (c == '.') {
		dstring_append_char(token, (char) c);
		state = OID;
	    } else {
		dstring_append_char(token, (char) c);
		state = TXT;
	    }
	    break;
	case OID:
	    if (isdigit(c)) {
		dstring_append_char(token, (char) c);
	    } else if (c == '.') {
		dstring_append_char(token, (char) c);
		state = OIDDOT;
	    } else {
		translate(token, subst);
		if (fFlag) {
		    if (dstring_len(subst) < dstring_len(token)) {
			dstring_expand(subst, dstring_len(token), ' ');
		    }
		}
		fputs(dstring_str(subst), stdout);
		if (dstring_len(subst) > dstring_len(token)) {
		    space = dstring_len(subst) - dstring_len(token) - 1;
		} else {
		    space = 0;
		}
		if (fFlag && space > 0 && c == ' ') {
		    state = EATSPACE;
		    space--;
		} else {
		    state = TXT;
		    space = 0;
		    fputc(c, stdout);
		}
		dstring_truncate(token, 0);
	    }
	    break;
	case OIDDOT:
	    if (isdigit(c)) {
		dstring_append_char(token, (char) c);
		state = OID;
	    } else {
		translate(token, subst);
		fputs(dstring_str(subst), stdout);
		fputc(c, stdout);
		dstring_truncate(token, 0);
		state = TXT;
	    }
	    break;
	case EATSPACE:
	    if (c == ' ' && space > 0) {
		space--;
	    } else {
		state = TXT;
	    }
	    break;
	}
    }

    if (dstring_len(token)) {
	switch (state) {
	case TXT:
	case NUM:
	case NUMDOT:
	case NUMDOTNUM:
	    fputs(dstring_str(token), stdout);
	    dstring_truncate(token, 0);
	    fputc(c, stdout);
	    fflush(stdout);
	    break;
	case OID:
	    if (isdigit(c)) {
		dstring_append_char(token, (char) c);
	    } else {
		translate(token, subst);
		if (fFlag) {
		    if (dstring_len(subst) < dstring_len(token)) {
			dstring_expand(subst, dstring_len(token), ' ');
		    }
		}
		fputs(dstring_str(subst), stdout);
		if (dstring_len(subst) > dstring_len(token)) {
		    space = dstring_len(subst) - dstring_len(token) - 1;
		} else {
		    space = 0;
		}
		if (fFlag && space > 0 && c == ' ') {
		    space--;
		} else {
		    space = 0;
		    fputc(c, stdout);
		}
		dstring_truncate(token, 0);
	    }
	    break;
	case OIDDOT:
	    if (isdigit(c)) {
		dstring_append_char(token, (char) c);
	    } else {
		translate(token, subst);
		fputs(dstring_str(subst), stdout);
		fputc(c, stdout);
		dstring_truncate(token, 0);
	    }
	    break;
	case EATSPACE:
	    if (c == ' ' && space > 0) {
		space--;
	    } else {
		fputc(c, stdout);
	    }
	    break;
	}
    }
}



static void usage()
{
    fprintf(stderr,
	    "Usage: smixlate [options] [module or path ...]\n"
	    "  -V, --version         show version and license information\n"
	    "  -h, --help            show usage information\n"
	    "  -c, --config=file     load a specific configuration file\n"
	    "  -p, --preload=module  preload <module>\n"
	    "  -r, --recursive       print errors also for imported modules\n"
	    "  -l, --level=level     set maximum level of errors and warnings\n"
	    "  -i, --ignore=prefix   ignore errors matching prefix pattern\n"
	    "  -I, --noignore=prefix do not ignore errors matching prefix pattern\n"
	    "  -a, --all             replace all OIDs (including OID prefixes)\n"
	    "  -f, --format          preserve formatting as much as possible\n");
}



static void help() { usage(); exit(0); }
static void version() { printf("smixlate " SMI_VERSION_STRING "\n"); exit(0); }
static void config(char *filename) { smiReadConfig(filename, "smixlate"); }
static void preload(char *module) { smiLoadModule(module); }
static void recursive() { flags |= SMI_FLAG_RECURSIVE; smiSetFlags(flags); }
static void level(int lev) { smiSetErrorLevel(lev); }
static void ignore(char *ign) { smiSetSeverity(ign, 128); }
static void noignore(char *ign) { smiSetSeverity(ign, -1); }



int main(int argc, char *argv[])
{
    int i;

    static optStruct opt[] = {
	/* short long              type        var/func       special       */
	{ 'a', "all",		 OPT_FLAG,   &aFlag,        0 },
	{ 'f', "format",         OPT_FLAG,   &fFlag,	    0 },
	{ 'h', "help",           OPT_FLAG,   help,          OPT_CALLFUNC },
	{ 'V', "version",        OPT_FLAG,   version,       OPT_CALLFUNC },
	{ 'c', "config",         OPT_STRING, config,        OPT_CALLFUNC },
	{ 'p', "preload",        OPT_STRING, preload,       OPT_CALLFUNC },
	{ 'r', "recursive",      OPT_FLAG,   recursive,     OPT_CALLFUNC },
	{ 'l', "level",          OPT_INT,    level,         OPT_CALLFUNC },
	{ 'i', "ignore",         OPT_STRING, ignore,        OPT_CALLFUNC },
	{ 'I', "noignore",       OPT_STRING, noignore,      OPT_CALLFUNC },
	{ 0, 0, OPT_END, 0, 0 }  /* no more options */
    };
    
    for (i = 1; i < argc; i++)
	if ((strstr(argv[i], "-c") == argv[i]) ||
	    (strstr(argv[i], "--config") == argv[i])) break;
    if (i == argc) 
	smiInit("smixlate");
    else
	smiInit(NULL);

    flags = smiGetFlags();
    flags |= SMI_FLAG_ERRORS;
    flags |= SMI_FLAG_NODESCR;
    smiSetFlags(flags);

    optParseOptions(&argc, argv, opt, 0);

    for (i = 1; i < argc; i++) {
	if (smiLoadModule(argv[i]) == NULL) {
	    fprintf(stderr, "smixlate: cannot locate module `%s'\n",
		    argv[i]);
	    smiExit();
	    exit(1);
	}
    }


    process(stdin);

    smiExit();

    return 0;
}
